### ISSUE TITLE
Ray does not support cython?

### ISSUE THREAD

**Issue by maxim28** [ray-project/ray#944]
i wrote a simple cython file which only has one actor with __init__ function, build and setup it.
```
@ray.remote
class SelfPlayer(object):
    def __init__(self):
        os.environ["CUDA_VISIBLE_DEVICES"] = ",".join([str(i) for i in ray.get_gpu_ids()])
```

In another python file, i import that module and try to initialize the SelfPlayer with ray.remote
`rewards = [selfplayers[i].do_selfplay.remote(args) for i in xrange(n_gpu)]`

then i got the following error
```
Traceback (most recent call last):
  File "policy/rlpolicy_train.py", line 218, in <module>
    train(args)
  File "policy/rlpolicy_train.py", line 57, in train
    selfplayers = [SelfPlayer.remote() for _ in xrange(n_gpu)]
  File "/usr/local/lib/python2.7/dist-packages/ray/actor.py", line 415, in remote
    actor_object._manual_init(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/ray/actor.py", line 434, in _manual_init
    signature.check_signature_supported(v, warn=True)
  File "/usr/local/lib/python2.7/dist-packages/ray/signature.py", line 47, in check_signature_supported
    in funcsigs.signature(func).parameters.items()]
  File "/usr/local/lib/python2.7/dist-packages/funcsigs/__init__.py", line 62, in signature
    sig = signature(obj.__func__)
  File "/usr/local/lib/python2.7/dist-packages/funcsigs/__init__.py", line 187, in signature
    raise ValueError('callable {0!r} is not supported by signature'.format(obj))
ValueError: callable <cyfunction SelfPlayer.__init__ at 0x7fb0c9ee1950> is not supported by signature
```

**Comment by pcmoritz** [comment]
Hey Maxim,

Supporting Cython actors would be a really cool feature to have and we would like to support it. Let us know if you are interested in looking into which changes are required to get it to work!

-- Philipp.

**Comment by danielsuo** [comment]
@mihaic and I are taking a look at this for the [BrainIAK](https://github.com/brainiak/brainiak) and DeepDrive projects. You can see our progress from poking around yesterday [here](https://github.com/danielsuo/ray/pull/1).

We abused ray to get a simple example working. Based on this experience, we think that supporting Cython actors / functions is doable, but may be a little finicky and limited in some ways (tbd).

We'll keep poking and come back with a fuller picture of limitations / proposed implementation for you guys to review.

**Comment by robertnishihara** [comment]
Nice, that's really awesome! Let us know what you find or if you run into any serious obstacles.

**Comment by danielsuo** [comment]
Just had another look (you can thank toddler nightmares) and was able to get both functions and actors working. I haven't made a pull request to this repository yet because I (sheepishly) admit I don't know what the conventions are around these parts. See pull request to my own fork [here](https://github.com/danielsuo/ray/pull/1).

Please let me know what testing / process I should go through (or if there's a CONTRIBUTING doc I missed) to move things along! This feature should be considered experimental since we've only spot-checked the various things Cython can do.

A few notes:
1. I am of the growing opinion that Cython should be kept separate from Python code. We would need to do some pretty ugly things to @maxim28's example working, but calling ```ray.remote``` from outside the ```*.pyx``` file doesn't require the same gymnastics. See the [README](https://github.com/danielsuo/ray/blob/8135427ad38c45a3d5fd2a48abcd998d79966470/examples/cython/README.md) I put together
2. Along these lines, we don't do anything about distributing / pickling Cython code. We assume that your code has already been distributed. This isn't a great answer for folks actively developing a Cython module since compiling / deploying to a cluster would be annoying, but it's not yet clear what the right answer is. For example, do we deploy a single binary assuming all machines in the cluster can run the same binary or should each machine recompile once it receives source code?
3. The [changes](https://github.com/danielsuo/ray/pull/1/files) to the ```ray``` code base are (imho) pretty minimal with the glaring exception of "ripping" signatures from Cython functions. I'm not proud of it, but it appears to get the job done. There are other options like forking ```funcsigs```, but I'm not sure we want to do that since ```funcsigs``` is a backport of a well-documented feature. Welcome all suggestions!

**Comment by danielsuo** [comment]
@mihaic may have suggestions for how to distribute Cython code

**Comment by robertnishihara** [comment]
I agree let's not worry about distributing Cython code for now (we're already assuming that all Python modules that are used live on all the machines).

I'd suggest submitting a PR with the following:
- the relevant changes to ray
- a simple example under `examples/cython` as you've done
- also, run the cython example and verify that it ran correctly in our travis tests, it would make sense as one of the tests listed in https://github.com/ray-project/ray/blob/master/.travis.yml#L101

You can try running the tests in the Travis file by hand (e.g., `python test/runtest.py`). You can also enable Travis for your fork so that you can run tests there before creating the PR.

It's also fine to create the PR and just mention that it's a work in progress.

**Comment by robertnishihara** [comment]
@maxim28, this should be addressed now by #1193, can you give it a try?

**Comment by virtualluke** [comment]
Just a somewhat related question, what about numba support, similar to do or very different?

**Comment by robertnishihara** [comment]
I think it should be similar (relevant issue is #593).

**Comment by danielsuo** [comment]
Definitely related, but both easier and harder to integrate than Cython. Easier because we don't deal with compilation (not totally true; you can pre-compile), harder because there are a lot of new function types that we could support that don't appear to have a common base class that I'm aware of (e.g., ```@cuda.jit``` and ```@jit``` return objects that only have common ancestor ```object```). I'm by no means a numba expert though, so would be delighted to be corrected.

We could enumerate the various function types we want to support (maybe a catch-all ```is_ray_func``` and ```is_ray_method```), but @robertnishihara's workaround in #593 is similar to what we have to do for Cython anyway (though, admittedly, their use cases are a different with numba much more likely used in driver code).

If folks are into it, I could do this pretty quickly, but I'm not sure if 1) the workaround is painful enough and 2) enumerating function types (absent some better mechanism) is desirable.

**Comment by robertnishihara** [comment]
Closing because this was fixed in #1193.
